### ISSUE TITLE
Add support for unittest2.TestCase.assertRaises as a context manager to unittest.TestCase

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5339]
|<img alt="JB's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| JB reported|
|-|-|
|Trac ID|trac#5339|
|Type|enhancement|
|Created|2011-10-26 23:17:49Z|
|Branch|https://github.com/twisted/twisted/tree/assertRaises-context-manager-5339-2|

Starting in 2.7, the assertRaises on the standard library's unittest.TestCase can be used as a context manager. Out of all the changes done in unittest2, this is the one I personally use the most, since the code ends up nice and readable.

Attachments:

* [5339-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e78/e78a99552c0bfc596d69733d6b52fa9d9d36230c/8a222c90223ac7ea312eafd9222c8b408821a1e1.diff) (7073 bytes) - added by adiroiban on 2014-03-23 17:50:56Z - 
* [5339-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e78/e78a99552c0bfc596d69733d6b52fa9d9d36230c/8fe6ff6fab9ccd84551f509cd909bcdeee4f7298.diff) (7350 bytes) - added by adiroiban on 2014-04-15 11:00:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5339 5339
type__enhancement enhancement
reporter__JB JB
priority__normal normal
milestone__ 
branch__branches_assertRaises_context_manager_5339_2 branches/assertRaises-context-manager-5339-2
branch_author__exarkun__habnabit exarkun, habnabit
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1319671069000000 1319671069000000
changetime__1403730738918971 1403730738918971
version__None None
owner__exarkun exarkun
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

[#7040](https://github.com/twisted/twisted/issues/7040) is a duplicate of this one.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban removed owner|
|-|-|

I have implemented support for context manager.

It is based on code from cpython (https://github.com/python/cpython/blob/master/Lib/unittest/case.py).

The cpython code was heavily modified to be compatible with current assertRaises implementation and with Twisted coding standard.

Please advise how to document the methods. I tried my best.

Added test for new context manager mode. Please let me know if I forgot a test.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_5"></a>|@habnabit commented|
|-|-|

(In [42206]) Branching to assertRaises-context-manager-5339.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_6"></a>|@habnabit commented|
|-|-|

(In [42207]) Apply patch from adiroiban. refs [#5339](https://github.com/twisted/twisted/issues/5339)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit commented|
|-|-|

[Forced a build.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assertRaises-context-manager-5339)

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_8"></a>|@habnabit set owner to @adiroiban|
|-|-|

[Well, buildbot fails a lot.](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/assertRaises-context-manager-5339)

The code looks good, aside from that the tests need to be fixed. The topfile is a bit awkwardly-worded, though. I would say "twisted.trial.unittest.SynchronousTestCase.assertRaises now can act as a context manager." Also, the docs need @param and @returns annotations.

Thanks for the contribution, though!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban removed owner|
|-|-|

Hi, Thanks for the review.

I have no idea why the test fail on buildbot.

They pass on my local computer. Did they all fail on your local computer?

twistedchecker fails at some pep8 step... maybe it does not work with latest pep8 version... ugly stuff :(T

I have attached a new patch based on latest trunk with updated news files and documentation.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [42802]) Branching to 'assertRaises-context-manager-5339-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [42803]) Apply 5339-2.diff 

refs [#5339](https://github.com/twisted/twisted/issues/5339)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Thanks.  This looks pretty good.  I am going to make some small tweaks, mostly to documentation and naming, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [42805]) Some doc, naming, and code arrangement tweaks

refs [#5339](https://github.com/twisted/twisted/issues/5339)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> Thanks. This looks pretty good. I am going to make some small tweaks, mostly to documentation and naming, and then merge.

Ha ha.  Joke's on me.  There were almost as many Python 2.6 incompatibilities as there were lines in the change.  And to make up the difference, Python 3.3 incompatibility.  Hooray.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [42811]) Merge assertRaises-context-manager-5339-2

Author: adiroiban, exarkun
Reviewer: exarkun
Fixes: [#5339](https://github.com/twisted/twisted/issues/5339)

Make `twisted.trial.unittest.SynchronousTestCase.assertRaises` a context
manager when it is called with only one positional argument.

This makes it like the standard library unittest module's `assertRaises`.
